compiler/src/dotty/tools/dotc/transform/SuperPhase.scala

```diff
@@ -0,0 +1,625 @@
+          val self = transformSpecificTree(tree1.self, start)
+          val body = transformStats(tree1.body, tree1.symbol, start)
+          trans(cpy.Template(tree1)(constr, parents, self, body))
+        case Tag.PackageDef =>
+          val tree1 = tree.asInstanceOf[PackageDef]
+          def transformPackage(implicit ctx: Context) = {
+            val pid = transformSpecificTree(tree1.pid, start)
+            val stats = transformStats(tree1.stats, tree.symbol, start)
+            trans(cpy.PackageDef(tree1)(pid, stats))
+          }
+          transformPackage(localContext)
+        case Tag.Thicket =>
+          val tree1 = tree.asInstanceOf[Thicket]
+          cpy.Thicket(tree1)(transformTrees(tree1.trees, start))
+        case _ =>
+          trans(tree)
+      }
+    }
+  }
+
+  def transformSpecificTree[T <: Tree](tree: T, start: Int)(implicit ctx: Context): T =
+    transformTree(tree, start).asInstanceOf[T]
+
+  def transformStats(trees: List[Tree], exprOwner: Symbol, start: Int)(implicit ctx: Context): List[Tree] = {
+    def transformStat(stat: Tree)(implicit ctx: Context): Tree = stat match {
+      case _: Import | _: DefTree => transformTree(stat, start)
+      case Thicket(stats) => cpy.Thicket(stat)(stats.mapConserve(transformStat))
+      case _ => transformTree(stat, start)(ctx.exprContext(stat, exprOwner))
+    }
+    val nestedCtx = statsContextTransformer(start)(trees, ctx)
+    val newTrees = flatten(trees.mapConserve(transformStat(_)(nestedCtx)))
+    statsNodeTransformer(start)(newTrees, nestedCtx)
+  }
+
+  def transformUnit(tree: Tree)(implicit ctx: Context) = {
+    val nestedCtx = unitContextTransformer(0)(tree, ctx)
+    val newTree = transformTree(tree, 0)(nestedCtx)
+    unitNodeTransformer(0)(newTree, nestedCtx)
+  }
+
+  def transformTrees(trees: List[Tree], start: Int)(implicit ctx: Context): List[Tree] =
+    flatten(trees.mapConserve(transformTree(_, start)))
+
+  def transformSpecificTrees[T <: Tree](trees: List[T], start: Int)(implicit ctx: Context): List[T] =
+    transformTrees(trees, start).asInstanceOf[List[T]]
+
+  override def run(implicit ctx: Context): Unit =
+    ctx.compilationUnit.tpdTree = transformUnit(ctx.compilationUnit.tpdTree)
+
+  // The rest of this class is all made up by initialization code
+
+  for ((phase, idx) <- miniPhases.zipWithIndex) {
+    phase.superPhase = this
+    phase.idxInGroup = idx
+  }
+
+  /** Class#getDeclaredMethods is slow, so we cache its output */
+  private val clsMethodsCache = new java.util.IdentityHashMap[Class[_], Array[java.lang.reflect.Method]]
+
+  /** Does `phase` contain a redefinition of method `name`?
+   *  (which is a method of MiniPhase)
+   */
+  private def defines(phase: MiniPhase, name: String) = {
+    def hasRedefinedMethod(cls: Class[_]): Boolean =
+      if (cls.eq(classOf[MiniPhase])) false
+      else {
+        var clsMethods = clsMethodsCache.get(cls)
+        if (clsMethods eq null) {
+          clsMethods = cls.getDeclaredMethods
```

Comment:
proposal: add `map(_.getName)` here and use `contains` instead of `exists` below.